wandb/sdk/launch/builder/kaniko_builder.py

```diff
@@ -0,0 +1,378 @@
+    create_metadata_file,
+    get_entry_point_command,
+)
+from ..utils import LOG_PREFIX, get_kube_context_and_api_client, sanitize_wandb_api_key
+from .build import _create_docker_build_ctx, generate_dockerfile
+
+_DEFAULT_BUILD_TIMEOUT_SECS = 1800  # 30 minute build timeout
+
+
+def _wait_for_completion(
+    batch_client: client.BatchV1Api, job_name: str, deadline_secs: Optional[int] = None
+) -> bool:
+    start_time = time.time()
+    while True:
+        job = batch_client.read_namespaced_job_status(job_name, "wandb")
+        if job.status.succeeded is not None and job.status.succeeded >= 1:
+            return True
+        elif job.status.failed is not None and job.status.failed >= 1:
+            wandb.termerror(f"{LOG_PREFIX}Build job {job.status.failed} failed {job}")
+            return False
+        wandb.termlog(f"{LOG_PREFIX}Waiting for build job to complete...")
+        if deadline_secs is not None and time.time() - start_time > deadline_secs:
+            return False
+
+        time.sleep(5)
+
+
+class KanikoBuilder(AbstractBuilder):
+    """Builds a docker image for a project using Kaniko."""
+
+    type = "kaniko"
+
+    build_job_name: str
+    build_context_store: str
+    secret_name: Optional[str]
+    secret_key: Optional[str]
+
+    def __init__(
+        self,
+        environment: AbstractEnvironment,
+        registry: AbstractRegistry,
+        build_job_name: str = "wandb-launch-container-build",
+        build_context_store: str = "",
+        secret_name: str = "",
+        secret_key: str = "",
+        verify: bool = True,
+    ):
+        """Initialize a KanikoBuilder.
+
+        Args:
+            environment (AbstractEnvironment): The environment to use.
+            registry (AbstractRegistry): The registry to use.
+            build_job_name (str, optional): The name of the build job.
+            build_context_store (str, optional): The name of the build context store.
+            secret_name (str, optional): The name of the secret to use for the registry.
+            secret_key (str, optional): The key of the secret to use for the registry.
+            verify (bool, optional): Whether to verify the functionality of the builder.
+                Defaults to True.
+
+        Raises:
+            LaunchError: If the builder cannot be intialized or verified.
+        """
+        if build_context_store is None:
+            raise LaunchError(
+                "You are required to specify an external build "
+                "context store for Kaniko builds. Please specify a  storage url "
+                "in the 'build-context-store' field of your builder config."
+            )
+        self.environment = environment
+        self.registry = registry
+        self.build_job_name = build_job_name
+        self.build_context_store = build_context_store.rstrip("/")
+        self.secret_name = secret_name
+        self.secret_key = secret_key
+        if verify:
+            self.verify()
```

Comment:
Do we not need to check `login` like in the dockerBuilder here?

Reply:
I guess we probably don't have to log into kaniko lol

Reply:
yeah you can't really login, but assuming that the registry was properly verified the credentials it provides should be legit. The unused kwargs is a really annoying side effect of using abstract base classes and abstract methods to tighten the class structure.